### PR TITLE
[SPARK-32314][SHS] Add config to control whether log old format of stacktrace

### PR DESCRIPTION
**Pull request by warrenzhu25**
### What changes were proposed in this pull request?
Avoid writing old format of stacktrace to event log to save space

### Why are the changes needed?
Old format stacktrace info is duplicated and cause huge event log 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Manually tested.

### CODE REVIEW

**Comment by warrenzhu25** [comment]
@sarutak @gengliangwang could you help take a look?

**Comment by warrenzhu25** [comment]
@HeartSaVioR Could you help take a look?

**Comment by HeartSaVioR** [comment]
It might be only me, but even we are trying to guarantee backward/forward compatibility with event logs, I wouldn't think it's for covering two major versions, even not latest one for 1.x.

IMHO the config is not needed and we can just remove the compatibility, but again that's only me, and it'd be nice to initiate discussion thread in dev@ list to reach consensus though. Could you go through the dev@ mailing list? I guess the discussion thread won't hold too long, as it might be just obvious.

**Comment by sarutak** [comment]
@warrenzhu25 
Thank you for the contribution but I don't think the compatibility for 1.2.0 and before is needed too. Considering our versioning policy, it's too old to support.
http://spark.apache.org/versioning-policy.html

**Comment by warrenzhu25** [comment]
> @warrenzhu25
> Thank you for the contribution but I don't think the compatibility for 1.2.0 and before is needed too. Considering our versioning policy, it's too old to support.
> http://spark.apache.org/versioning-policy.html

Do you mean I can just remove the config?

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 1]
```diff
@@ -175,6 +175,14 @@ package object config {
       .booleanConf
       .createWithDefault(false)
 
+  private[spark] val EVENT_LOG_OLD_STACK_TRACE =
+    ConfigBuilder("spark.eventLog.logOldStackTrace")
+      .doc("Whether to write old format of stacktrace to the event log. " +
+        "Only enable this when spark history server is before version 1.2.0")
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:181]: Which version is being referred to here with `1.2.0` ?

Comment (warrenzhu25) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:181]: It's in `TaskEndReason`. The code is as below:
```
  override def toErrorString: String =
    if (fullStackTrace == null) {
      // fullStackTrace is added in 1.2.0
      // If fullStackTrace is null, use the old error string for backward compatibility
      exceptionString(className, description, stackTrace)
    } else {
      fullStackTrace
    }
```

**Comment by warrenzhu25** [comment]
@sarutak @gengliangwang I removed the config directly. Could you help take a look again?

**Comment by mridulm** [comment]
I agree with @HeartSaVioR - compatibility with 1.2.0 is not required.

**Comment by mridulm** [comment]
Briefly looking at spark code, this is currently still in use - looks like that comment in `TaskEndReason` might not be valid anymore.
See [here](https://github.com/apache/spark/blob/594c7c613a8ef80ab6b3725832579f12d40b02c8/core/src/main/scala/org/apache/spark/executor/Executor.scala#L694) for example.

There are other uses as well, given that I am not in favor of this change. If it had been a truely dead code, I would have been more inclined to cleanup (with a caveat that this is part of public api, and so we need to be careful with breaking behavior).

**Comment by warrenzhu25** [comment]
> Briefly looking at spark code, this is currently still in use - looks like that comment in `TaskEndReason` might not be valid anymore.
> See [here](https://github.com/apache/spark/blob/594c7c613a8ef80ab6b3725832579f12d40b02c8/core/src/main/scala/org/apache/spark/executor/Executor.scala#L694) for example.
> 
> There are other uses as well, given that I am not in favor of this change. If it had been a truely dead code, I would have been more inclined to cleanup (with a caveat that this is part of public api, and so we need to be careful with breaking behavior).

Thanks for comment. For this example, it's about how to create ExceptionFailure. This pr is only about when writing into event log, also affecting TaskEndReason after replay event. And there is no api directly use `stackTrace: Array[StackTraceElement]`, this only used in below:

```
  override def toErrorString: String =
    if (fullStackTrace == null) {
      // fullStackTrace is added in 1.2.0
      // If fullStackTrace is null, use the old error string for backward compatibility
      exceptionString(className, description, stackTrace)
    } else {
      fullStackTrace
    }
```

**Comment by mridulm** [comment]
To clarify what I meant:
1. `ExceptionFailure` is a public api.
2. History server is not the only consumer of event files.
3. There are code paths which are populating `stackTrace` in `ExceptionFailure`

Given all of these, setting stack trace to empty is backwardly incompatible against current spark versions, and breaks compatibility.

I would have been in favour of the change if (3) above was not the case.
Given all of these three conditions currently applying, this is a breaking change and I am not in favor of it.

**Comment by warrenzhu25** [comment]
> To clarify what I meant:
> 
> 1. `ExceptionFailure` is a public api.
> 2. History server is not the only consumer of event files.
> 3. There are code paths which are populating `stackTrace` in `ExceptionFailure`
> 
> Given all of these, setting stack trace to empty is backwardly incompatible against current spark versions, and breaks compatibility.
> 
> I would have been in favour of the change if (3) above was not the case.
> Given all of these three conditions currently applying, this is a breaking change and I am not in favor of it.

Could you show me one example code path which is populating `stackTrace` in `ExceptionFailure`?

**Pull request closed without merging**